xdem/spatial_tools.py

```diff
@@ -98,65 +98,142 @@ def merge_bounding_boxes(bounds: list[rio.coords.BoundingBox], resolution: float
         for key in "bottom", "left":
             max_bounds[key] = np.nanmin([max_bounds[key], bound.__getattribute__(key)])
 
-    for key in max_bounds:
-        modulo = max_bounds[key] % resolution
-        max_bounds[key] -= modulo
-
-        if key in ["right", "top"] and modulo > 0:
-            max_bounds[key] += resolution
+    # Make sure that extent is a multiple of resolution
+    for key1, key2 in zip(("left", "bottom"), ("right", "top")):
+        modulo = (max_bounds[key2] - max_bounds[key1]) % resolution
+        max_bounds[key2] += modulo
 
     return rio.coords.BoundingBox(**max_bounds)
 
 
-def merge_rasters(rasters: list[gu.georaster.Raster], reference: int = 0, merge_algorithm: Callable = np.nanmean,
-                  resampling_method: Union[str, rio.warp.Resampling] = "nearest") -> gu.georaster.Raster:
+def stack_rasters(rasters: list[gu.georaster.Raster], reference: Union[int, gu.Raster] = 0,
+                  resampling_method: Union[str, rio.warp.Resampling] = "bilinear",
+                  use_ref_bounds = False) -> tuple[gu.georaster.Raster, rio.coords.BoundingBox]:
     """
-    Merge a list of rasters into one larger raster.
+    Stack a list of rasters into a common grid as a 3D np array with nodata set to Nan.
 
```

Comment:
Could you write the output shape here somewhere to make it easier to understand what it does? Something alike:
```
If `reference` is an index, the output will have a shape of (N, height, width) where N is len(rasters) - 1 and height and width is equal to rasters[0].shape.
If `reference` is another raster, N will be equal to len(rasters), and height and width will be equal to reference.shape .
```
Do I understand the shapes correctly?

Reply:
Actually, N will always be the same, as all items in rasters are now stacked (since reference is external). So there is no distinction between both.
Additionally, it's difficult to state what the output shape will be. If use_ref_bounds is True, then yes it will have the same dimensions, otherwise not as it will depend on the max bounds. Would something like this help?

> 
> If use_ref_bounds is True, output will have the shape (N, height, width) where N is len(rasters) and height and width is equal to reference's shape.
> If use_ref_bounds is False, output will have the shape (N, height2, width2) where N is len(rasters) and height2 and width2 are set based on reference's resolution and the maximum extent of all rasters.

Reply:
Wait, so if this is run:
```python
stacked = stack_rasters([ref, nonref0, nonref1], reference=0)
```
`stacked.shape[0]` will be 2??

Reply:
No, as stated in the new docstring, stacked.shape[0] is **always** equal to len(rasters). Now the way 'reference' is excluded is by setting it to a separate dataset rather than an index. Is that clear?

Reply:
I see!